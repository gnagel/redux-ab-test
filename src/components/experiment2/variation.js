import React                  from 'react';
import { connect }            from 'react-redux';


class Variation extends React.Component {
  render() {
    const { children } = this.props;
    return children;
  }
}


Variation.propTypes = {
  // Name of your variation
  name:       React.PropTypes.string.isRequired,
  // React element to render:
  children:   React.PropTypes.node.isRequired,
};


const mapStateToProps = ({ reduxAbTest }) => ({ reduxAbTest });
export default connect(mapStateToProps, null)(Variation);
