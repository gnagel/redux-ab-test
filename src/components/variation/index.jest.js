import Variation from './index';

describe('(Component) src/components/variation/index.js', () => {
  it('Variation exists', () => {
    expect(Variation).not.toBeUndefined;
  });
});
