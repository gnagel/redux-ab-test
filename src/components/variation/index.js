import React     from 'react';
import Variation from './variation';
export default Variation;
export const isVariation = (child) => (React.isValidElement(child) && child.type == Variation);
