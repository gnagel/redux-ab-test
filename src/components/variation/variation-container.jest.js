import React from 'react'; // eslint-disable-line no-unused-vars
import Immutable from 'immutable';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Variation, { mapStateToProps } from './variation';
import reduxAbTest, { initialState } from '../../module';


describe('(Container) Variation', () => {
  describe('mapStateToProps', () => {
    it('exists', () => {
      expect(mapStateToProps).not.toBeUndefined;
    });
    it('has correct keys', () => {
      expect(Object.keys(mapStateToProps({reduxAbTest: 'test-state'}))).toEqual(['reduxAbTest']);
    });
    it('has correct state', () => {
      expect(mapStateToProps({reduxAbTest: 'test-state'})).toEqual({reduxAbTest: 'test-state'});
    });
  });

  describe('component', () => {
    let component;
    let tree;
    beforeEach(() => {
      const props = {
        name:       'Variation B',
        experiment: Immutable.fromJS({
          name:       'Test-experimentName',
          variations: [
            { name: 'Original',  weight: 10000 },
            { name: 'Variation B', weight: 0 },
          ],
        }),
        children: 'Test Variation B',
      };
      const state = {
        reduxAbTest: initialState.merge({
          experiments: [
            {
              name:       'Test-experimentName',
              variations: [
                { name: 'Original', weight: 10000 },
                { name: 'Variation B', weight: 0 },
              ],
            },
          ],
        }),
      };
      component = renderer.create(
        <Provider store={createStore(reduxAbTest, state)}>
          <Variation {...props} />
        </Provider>
      );
      tree = component.toJSON();
    });

    it('variation has expected props', () => {
      expect(tree).toMatchSnapshot();
      expect(tree.type).toEqual('span');
      expect(tree.props['data-experiment-id']).toBeUndefined;
      expect(tree.props['data-experiment-name']).toEqual('Test-experimentName');
      expect(tree.props['data-variation-id']).toBeUndefined;
      expect(tree.props['data-variation-name']).toEqual('Variation B');
      expect(tree.children).toEqual(['Test Variation B']);
    });
  });
});
