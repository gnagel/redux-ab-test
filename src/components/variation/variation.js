/** @flow */
import React from 'react';
import Immutable from 'immutable';

type Props = {
  id:         string,
  name:       string,
  experiment: Immutable.Map,
  handleWin:  Function,
};

export default class Variation extends React.Component {
  props: Props;
  static defaultProps = {
    id:         null,
    name:       null,
    experiment: Immutable.Map({}),
    handleWin:  () => {}
  };

  render() {
    const { id, name, experiment, handleWin, children } = this.props;

    const childrenProps = {
      'data-experiment-id':   experiment.get('id'),
      'data-experiment-name': experiment.get('name'),
      'data-variation-id':    id,
      'data-variation-name':  name,
    };

    // This is text or null content, wrap it in a span and return the contents
    if (!React.isValidElement(children)) {
      return (
        <span
          data-experiment-id={experiment.get('id')}
          data-experiment-name={experiment.get('name')}
          data-variation-id={id}
          data-variation-name={name}>
          {children}
        </span>
      );
    }

    // Inject the experiment/variation props into the children
    return React.cloneElement(children, {
      variationId:    id,
      variationName:  name,
      experimentId:   experiment.get('id'),
      experimentName: experiment.get('name'),
      handleWin
    });
  }
}
