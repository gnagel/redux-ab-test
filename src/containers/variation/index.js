import Variation from './variation';
export default Variation;
