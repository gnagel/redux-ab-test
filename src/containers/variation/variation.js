/** @flow */
import React from "react"; // eslint-disable-line no-unused-vars
import { connect } from 'react-redux';
import Variation from '../../components/variation';


/**
 * Map the Redux Store to the Experiment's props
 */
export const mapStateToProps = (state:Object) => {
  const { reduxAbTest } = state;
  return { reduxAbTest };
};


/**
 * Export the new React Container.
 */
export default connect(mapStateToProps)(Variation);
