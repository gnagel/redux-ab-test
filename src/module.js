/** @flow */
import React                           from "react"; // eslint-disable-line no-unused-vars
import Immutable                       from 'immutable';
import { createAction, handleActions } from 'redux-actions';
import findExperiment                  from './utils/find-experiment';
import randomVariation                 from './utils/random-variation';
import selectVariation                 from './utils/select-variation';
import createCacheStore                from './utils/create-cache-store';

export const cacheStore = createCacheStore();

export type VariationType = {
  id: ?string,
  name: string,
  weight: number,
};

export type ExperimentType = {
  id: ?string,
  name: string,
  variations: Array<VariationType>,
};


export const constants = {
  RESET: 'redux-ab-test/RESET',
  LOAD: 'redux-ab-test/LOAD',
  ACTIVATE: 'redux-ab-test/ACTIVATE',
  DEACTIVATE: 'redux-ab-test/DEACTIVATE',
  PLAY: 'redux-ab-test/PLAY',
  WIN: 'redux-ab-test/WIN'
};


export const actions = {
  reset: createAction(constants.RESET, () => Immutable.fromJS({})),
  load: createAction(constants.LOAD, ({experiments, active}) => {
    return Immutable.fromJS({experiments, active});
  }),
  activate: createAction(constants.ACTIVATE, ({experiment}) => {
    return Immutable.fromJS({experiment});
  }),
  deactivate: createAction(constants.DEACTIVATE, ({experiment}) => {
    return Immutable.fromJS({experiment});
  }),
  play: createAction(constants.PLAY, ({experiment, variation}) => {
    return Immutable.fromJS({experiment, variation});
  }),
  win: createAction(constants.WIN, ({experiment, variation}) => {
    return Immutable.fromJS({experiment, variation});
  })
};


export const initialState = Immutable.fromJS({
  experiments: [ /** Array of ExperimentType objects */ ],
  running: { /** "experiment name" => number */ },
  active: { /** "experiment name" => "variation name" */ },
  winners: { /** "experiment name" => "variation name" */ }
});


const reducers = {
  /**
   * RESET the experiments state.
   */
  [constants.RESET]: (state, { }) => {
    cacheStore.clear();
    return initialState;
  },

  /**
   * LOAD the available experiments. and reset the state of the server
   */
  [constants.LOAD]: (state, { payload }) => {
    return initialState.set('experiments', payload.get('experiments')).set('active', payload.get('active'));
  },

  /**
   * ACTIVATE the available experiments. and reset the state of the server
   */
  [constants.ACTIVATE]: (state, { payload }) => {
    const experimentName = payload.get('experiment').get('name');
    const counter = (state.get('running').get(experimentName) || 0) + 1;
    const running = state.get('running').set(experimentName, counter);
    return state.set('running', running);
  },

  /**
   * DEACTIVATE the available experiments. and reset the state of the server
   */
  [constants.DEACTIVATE]: (state, { payload }) => {
    const experimentName = payload.get('experiment').get('name');
    const counter = (state.get('running').get(experimentName) || 0) - 1;
    let running;
    if (counter <= 0) {
      running = state.get('running').delete(experimentName);
    } else {
      running = state.get('running').set(experimentName, counter);
    }
    return state.set('running', running);
  },

  /**
   * A user saw an experiment
   * @payload { experiment:ExperimentType, variation:VariationType }
   */
  [constants.PLAY]: (state, { payload }) => {
    const experimentName = payload.get('experiment').get('name');
    const variationName = payload.get('variation').get('name');
    const active = state.get('active').set(experimentName, variationName);
    cacheStore().removeItem(experimentName);
    return state.set('active', active);
  },

  /**
   * A user interacted with the variation
   * @payload { experiment:ExperimentType, variation:VariationType }
   */
  [constants.WIN]: (state, { payload }) => {
    const experimentName = payload.get('experiment').get('name');
    const variationName = payload.get('variation').get('name');
    const winners = state.get('winners').set(experimentName, variationName);
    return state.set('winners', winners);
  }
};

export default handleActions(reducers, initialState);



//
// Selectors for querying the Redux/Compoenet state:
//
export const selectors = {
  /**
   * Find the experiment by name, raises an Error if not-found
   */
  findExperiment,
  /**
   * Select a variation from the given input variables
   */
  selectVariation
};
