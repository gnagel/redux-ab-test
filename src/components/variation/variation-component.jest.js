import React from 'react'; // eslint-disable-line no-unused-vars
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Variation } from './variation';
import { initialState } from '../../module';



describe('(Component) src/components/variation/variation.js', () => {
  let shallowComponent;
  let component;
  let tree;
  let props;
  beforeEach(() => {
    const reduxAbTest = initialState.merge({
      'experiments': [
        {
          id:         'test-experimentId',
          name:       'Test-experimentName',
          variations: [
            { id: 'test-id-original',    name: 'Original',    weight: 10000 },
            { id: 'test-id-variation-b', name: 'Variation B', weight: 0 },
          ],
        },
      ],
      'active': { 'Test-experimentName': 'Variation B' },
    });

    props = {
      name:       'Variation B',
      experiment: reduxAbTest.getIn(['experiments', 0]),
      children:   'Test-children',
      reduxAbTest,
    };
    shallowComponent = shallow(
      <Variation {...props} />
    );
    component = renderer.create(
      <Variation {...props} />
    );
    tree = component.toJSON();
  });

  it('exists', () => {
    expect(component).not.toBeUndefined;
  });

  it('has the correct props', () => {
    expect(tree).toMatchSnapshot();
    expect(tree.props['data-experiment-id']).toEqual('test-experimentId');
    expect(tree.props['data-experiment-name']).toEqual('Test-experimentName');
    expect(tree.props['data-variation-id']).toEqual('test-id-variation-b');
    expect(tree.props['data-variation-name']).toEqual('Variation B');
    expect(tree.children).toEqual(['Test-children']);
  });

  it('has the correct text', () => {
    expect(component).to.have.text('Test-children');
  });

  it('has the correct tagName', () => {
    expect(shallowComponent).to.have.tagName('span');
  });

  it('wraps the text in a span', () => {
    const child = component.find('span');
    expect(child).to.have.tagName('span');
    expect(child).to.have.text('Test-children');
  });

  it('have the correct data-* props', () => {
    const child = shallowComponent.find('span');
    expect(child).to.have.prop('data-variation-id', 'test-id-variation-b');
    expect(child).to.have.prop('data-variation-name', 'Variation B');
    expect(child).to.have.prop('data-experiment-id', 'test-experimentId');
    expect(child).to.have.prop('data-experiment-name', 'Test-experimentName');
  });

  it('has the correct data-* attrs', () => {
    const child = shallowComponent.find('span');
    expect(child).to.have.data('variation-id', 'test-id-variation-b');
    expect(child).to.have.data('variation-name', 'Variation B');
    expect(child).to.have.data('experiment-id', 'test-experimentId');
    expect(child).to.have.data('experiment-name', 'Test-experimentName');
  });

  it('adds the props to the children component', () => {
    props = { ...props, children: (<div id="test-id">Test-children</div>) };
    component = shallow(
      <Variation {...props} />
    );
    expect(component).to.have.tagName('div');
    expect(component).to.have.attr('id', 'test-id');
    expect(component).to.have.data('variation-id', 'test-id-variation-b');
    expect(component).to.have.data('variation-name', 'Variation B');
    expect(component).to.have.data('experiment-id', 'test-experimentId');
    expect(component).to.have.data('experiment-name', 'Test-experimentName');
    expect(component).to.have.text('Test-children');
  });
});
