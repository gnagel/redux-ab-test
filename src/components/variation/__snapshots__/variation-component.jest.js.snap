// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) src/components/variation/variation.js has the correct props 1`] = `
<span
  data-experiment-id="test-experimentId"
  data-experiment-name="Test-experimentName"
  data-variation-id="test-id-variation-b"
  data-variation-name="Variation B"
>
  Test-children
</span>
`;
